### PR TITLE
Demand forecasting using TCN

### PR DESCRIPTION
**Pull request by vlbejan**
# Description
This pull request is for 2 notebooks. One for the demand forecast using TCN and the other one for the data prep.

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/Azure/azureml-examples/blob/main/CONTRIBUTING.md)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb [hunk 1]
```diff
@@ -0,0 +1,1471 @@
...
+    "\n",
+    "Having a problem description such as to generate accurate forecasts 24 hours ahead sounds like a relatively straght forward task. However, there are quite a few steps a user needs to take before the model is put in production. A user needs to prepare the data, partition it into appropriate sets, select the best model, evaluate it against a baseline, and monitor the model in real life to collect enough observations on how it would perform had it been put in production. Some of these steps are time consuming, some require certain expertise in writing code. The steps shown in this notebook follow a typical thought process one follows before the model is put in production.\n",
```
[the remark this answers is not in the document]
Comment (vlbejan) [reply v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb:43]: Thank you.

File: v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb [hunk 2]
```diff
@@ -0,0 +1,1471 @@
...
+    "\n",
+    "Make sure you have executed the [configuration](https://github.com/Azure/azureml-examples/blob/b13d6f9c5cf5c042bcc265fdbc4ed741c4527ded/v1/python-sdk/configuration.ipynb) before running this notebook."
```
[the remark this answers is not in the document]
Comment (vlbejan) [reply v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb:45]: Good catch. This link is broken for all notebooks. Will resolve the comments after updating the link. Need to find it first.

File: v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb [hunk 3]
```diff
@@ -0,0 +1,1471 @@
...
+    "auth = InteractiveLoginAuthentication(tenant_id = 'mytenantid')\n",
+    "ws = Workspace.from_config(auth = auth)\n",
+    "```\n",
+    "If you need to run in an environment where interactive login is not possible, you can use Service Principal authentication by replacing the ws = Workspace.from_config() line in the cell below with the following:\n",
+    "```\n",
+    "from azureml.core.authentication import ServicePrincipalAuthentication\n",
+    "auth = ServicePrincipalAuthentication('mytenantid', 'myappid', 'mypassword')\n",
+    "ws = Workspace.from_config(auth = auth)\n",
+    "```\n",
+    "For more details, see aka.ms/aml-notebook-auth"
```
[the remark this answers is not in the document]
Comment (vlbejan) [reply v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb:126]: Done

File: v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb [hunk 4]
```diff
@@ -0,0 +1,1471 @@
...
+    "\n",
+    "The `PythonScriptStep` uses the following arguments:<ul>\n",
+    "    <li> `model_name`: the name under which the model was registered in the workspace. </li>\n",
+    "    <li> `ouput_dataset_name`: name of the dataset under which the test set predictions will registered in the datastore</li>\n",
+    "    <li> `test_dataset_name`: name of the test dataset </li>\n",
+    "    <li> `target_column_name`: target column name</li>\n",
+    "    <li> `time_column_name`: time column name</li>\n",
+    "    <li> `output_path`: output data that is obtained from prediction step. </li>\n",
+    "    <li> `run_rolling_evaluation`: Boolean parameter. Set to True to run rolling evaluation. Otherwise, runs a regular forecast for the entire test set. </li>\n",
+    "    <li> `rolling_evaluation_step_size`: Optional parameter that instructs the rolling forecaster how many periods to step forward. See the <u>*Evaluating model accuracy with a rolling forecast*</u> section in this [document](https://learn.microsoft.com/en-us/azure/machine-learning/v1/how-to-auto-train-forecast-v1) for a more detailed explanation. </li></ul>\n",
```
[the remark this answers is not in the document]
Comment (vlbejan) [reply v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb:824]: Great suggestion. Thank you.

File: v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb [hunk 5]
```diff
@@ -0,0 +1,1471 @@
...
+    "In this section we will evaluate the test set performance for the best TCN model and compare it with the baseline. We will generate time series plots for forecasts and actuals, calculate accuracy metrics and plot the evolution of metrics for each model over time. All output from this section will be stored in the `forecast_output` folder and can be referenced any time you need it."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 7.1 Load test set results\n",
+    "\n",
+    "Here, we will import test set results for both many-models and baseline experiments."
```
[the remark this answers is not in the document]
Comment (vlbejan) [reply v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb:980]: Good catch.

File: v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (vlbejan) [reply v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb]: Done.

[the remark this answers is not in the document]
Comment (iamrk04) [reply v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb:45]: looks like the last commit overwrote the actual link:
https://github.com/Azure/azureml-examples/blob/main/sdk/python/jobs/configuration.ipynb

File: v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb (hunk 3 above)
[the remark this answers is not in the document]
Comment (iamrk04) [reply v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb]: This is now leading to Page not found error at GitHub.

File: v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb [hunk 6]
```diff
@@ -0,0 +1,1471 @@
...
+    "run_config.environment = inference_env"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 6.1.2 Build and submit the inference pipeline\n",
+    "\n",
+    "The inference pipeline will create two different output formats, 1) a tabular dataset that contains the prediction and 2) an `OutputFileDatasetConfig` that can be used for the sequential pipeline steps. The `inference_script_tcn.py` performs a rolling evaluation on the test set with the step size of 24 hours. This generates the same results as if the model was put in production and generated a 24-hour forecast once a day using the most recent data context. This is an efficient way of testing historical model performance at a tiny fraction of the compute and time commitment costs. For a more detailed information on rolling evaluation see the <u>*Evaluating model accuracy with a rolling forecast*</u> section [here](https://learn.microsoft.com/en-us/azure/machine-learning/v1/how-to-auto-train-forecast-v1).\n",
```
[the remark this answers is not in the document]
Comment (iamrk04) [reply v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb:812]: looks like this is still unchanged.

File: v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb [hunk 7]
```diff
@@ -0,0 +1,1470 @@
...
+    "auth = InteractiveLoginAuthentication(tenant_id = 'mytenantid')\n",
+    "ws = Workspace.from_config(auth = auth)\n",
+    "```\n",
+    "If you need to run in an environment where interactive login is not possible, you can use Service Principal authentication by replacing the ws = Workspace.from_config() line in the cell below with the following:\n",
+    "```\n",
+    "from azureml.core.authentication import ServicePrincipalAuthentication\n",
+    "auth = ServicePrincipalAuthentication('mytenantid', 'myappid', 'mypassword')\n",
+    "ws = Workspace.from_config(auth = auth)\n",
+    "```\n",
+    "For more details, see [this link](aka.ms/aml-notebook-auth)."
```
Comment (ctian-msft) [inline v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb:126]: Doesn't seem that customer outside Microsoft could directly use aka.ms link? Landed on a 404 page when clicking it directly. But, if I copy the link in the browser, it gives me the link that should be accessible by customers: https://github.com/Azure/MachineLearningNotebooks/blob/master/how-to-use-azureml/manage-azureml-service/authentication-in-azureml/authentication-in-azureml.ipynb

File: v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb [hunk 8]
```diff
@@ -0,0 +1,1470 @@
...
+    "    task=\"forecasting\",\n",
+    "    debug_log=\"baseline.log\",\n",
+    "    primary_metric=\"normalized_root_mean_squared_error\",\n",
+    "    experiment_timeout_hours=1,\n",
+    "    iterations=1,\n",
+    "    training_data=train_dataset,\n",
+    "    label_column_name=target_column_name,\n",
+    "    compute_target=compute_target,\n",
+    "    enable_early_stopping=True,\n",
+    "    n_cross_validations=2,\n",
```
Comment (ctian-msft) [inline v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb:716]: Shall we make a note that we're not actually using cross-validations in TCN?

File: v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (vlbejan) [reply v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb:45]: Correct. It will be fixed by the next commit.

File: v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb (hunk 3 above)
[the remark this answers is not in the document]
Comment (vlbejan) [reply v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb]: Changed to an actual link: https://github.com/Azure/MachineLearningNotebooks/blob/master/how-to-use-azureml/manage-azureml-service/authentication-in-azureml/authentication-in-azureml.ipynb

File: v1/python-sdk/tutorials/automl-with-azureml/forecasting-data-preparation/auto-ml-forecasting-data-preparation.ipynb [hunk 9]
```diff
@@ -0,0 +1,919 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 1. Introduction\n",
+    "\n",
+    "The objective of this notebook is to illustrate how to pre-process the raw data and register partitioned datasets to be used in demand forecasting notebooks: (i) Demand Forecastsing Using TCN ([link placeholder]()) and (ii) Demand Forecasting Using Many Models ([link placeholder]()).\n",
+    "\n",
+    "For illustration purposes we use the UCI electricity data ([link](https://archive.ics.uci.edu/ml/datasets/ElectricityLoadDiagrams20112014#)) which contains electricity consumption data for 370 consumers measured at 15 minute intervals. In this notebook, we will show how to ingest the data from the original source, aggregate to an hourly frequency, select a subsample of unique time series, determine the approriate way to partition the data, and finally, register the datasets to be used on the aforementoned notebooks.\n",
+    "\n",
+    "Make sure you have executed the [configuration](https://github.com/Azure/azureml-examples/blob/b13d6f9c5cf5c042bcc265fdbc4ed741c4527ded/v1/python-sdk/configuration.ipynb) before running this notebook."
```
[the remark this answers is not in the document]
Comment (vlbejan) [reply v1/python-sdk/tutorials/automl-with-azureml/forecasting-data-preparation/auto-ml-forecasting-data-preparation.ipynb:37]: Fixed.

File: v1/python-sdk/tutorials/automl-with-azureml/forecasting-data-preparation/auto-ml-forecasting-data-preparation.ipynb [hunk 10]
```diff
@@ -0,0 +1,919 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 3. Data work\n",
+    "We start with declaring some of the parameters that will be used in this notebook.\n",
+    "\n",
+    "- `IS_ORIGINAL_DATASET_DOWNLOADED` is a flag for wether we want to doanload the original data from the source. The flag is here to reduce the download time since the original dataset is larger than 1 GB.\n",
```
[the remark this answers is not in the document]
Comment (vlbejan) [reply v1/python-sdk/tutorials/automl-with-azureml/forecasting-data-preparation/auto-ml-forecasting-data-preparation.ipynb:115]: Done.

File: v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb (hunk 7 above)
Comment (vlbejan) [reply v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb:126]: Fixed.

File: v1/python-sdk/tutorials/automl-with-azureml/forecasting-data-preparation/auto-ml-forecasting-data-preparation.ipynb [hunk 11]
```diff
@@ -0,0 +1,919 @@
...
+   "source": [
+    "raw_df.rename(columns={raw_df.columns[0]: \"date\"}, inplace=True)\n",
+    "raw_df.head()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The \"wide\" data format is not usable in AutoML, which is deisgned to accept a \"long\" format data. To tranform the data from wide to long format, we will take the each uniqe time seres (date, MT_xxx) and stack them vertically. The end result will be a data frame containing 3 columns: (i) the `date` column, (ii) `custmer_id` column, which is the identifier of the time series and is derived from the column name in the wide format, and (iii) `usage` which is the target varaible we are trying to model."
```
[the remark this answers is not in the document]
Comment (vlbejan) [reply v1/python-sdk/tutorials/automl-with-azureml/forecasting-data-preparation/auto-ml-forecasting-data-preparation.ipynb:227]: Good point. I added a few sentences on the difference between configs for the baseline and TCN models.

File: v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb (hunk 8 above)
Comment (vlbejan) [reply v1/python-sdk/tutorials/automl-with-azureml/forecasting-demand-forecasting-tcn/auto-ml-forecasting-demand-tcn.ipynb:718]: Good point. I added a few sentences on the difference between configs for the baseline and TCN models.

**Comment by vlbejan** [comment]
After rebase all sorts of commits appeared. Will create a new PR. This will not be an issue since all comments are resolved.

**Comment by terrychang1015** [comment]
Hi Sir,

I lost my password for this Github account, could you help me to
unsubscribe this mail thread?

Regards
Terry
